### DISCUSSION TITLE
Import custom browser flow to new realm

### DISCUSSION THREAD

**Discussion by Bhavik237** [keycloak/keycloak#21396] [Q&A]
Hello everyone,
I have a custom browser flow for realm that I would like to apply whenever I create a new realm.
I have try exporting and importing realm but not able to get custom browser flow,
can you just me , what should be my approach? 

`curl --location 'http://127.0.0.1:8081/auth/admin/realms/jsontest/authentication/flows' \
--header 'Content-Type: application/json' \
--header 'Authorization: Bearer eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJQTl9NUlhpZnN5OGp3QnFuSi1hVVU0VnplMUUwM0hNUmhya0dZMXFMNWVVIn0.eyJleHAiOjE2ODgzNzkyMTIsImlhdCI6MTY4ODM3NzQxMiwianRpIjoiNjYzOGJkZWMtYzllNi00NDljLTk4ZDAtOWY3MDk3ODljMGQyIiwiaXNzIjoiaHR0cDovL2xvY2FsaG9zdDo4MDgxL2F1dGgvcmVhbG1zL21hc3RlciIsInN1YiI6IjNjNDQ3OTUyLTY5N2EtNDcwZi1hNjcwLTZjOWM0ODRmYzNmOCIsInR5cCI6IkJlYXJlciIsImF6cCI6ImFkbWluLWNsaSIsInNlc3Npb25fc3RhdGUiOiI3ZGRhZjFkYi1mOWM0LTQ4N2QtYWMwNS05OTIzMzA5OThhNmMiLCJhY3IiOiIxIiwic2NvcGUiOiJwcm9maWxlIGVtYWlsIiwic2lkIjoiN2RkYWYxZGItZjljNC00ODdkLWFjMDUtOTkyMzMwOTk4YTZjIiwiZW1haWxfdmVyaWZpZWQiOmZhbHNlLCJwcmVmZXJyZWRfdXNlcm5hbWUiOiJhZG1pbiIsImdpdmVuX25hbWUiOiIiLCJmYW1pbHlfbmFtZSI6IiIsImVtYWlsIjoiYmhhdmlrZHBhdGVsMjM3QGdtYWlsLmNvbSJ9.Y-8xSswkR1hgemXy9O4s9TF1LGdFrzTkgMdaOq_y5ZABZ3WWeWPnCCC1YG2og9H7C7IRCPWukqgpUiO_Ay5kPLpke6sF6WJilXA5ayGZiQCbLYeM-AtwkKBXfS-Re1ZA1WezQId_kGX-P3O4_1iVn64VOPA-lpnR55-Htp4PlJJVYs9x822Cnt-yho6ZVwKE1R9f9e_i7AlHI-QckrsE4_ivqOvRjx7j8XLjn7bKHVnw-c_0e22jgPLFe6NqOcaWZRTtvZu7wGIUjA-6heW9KgX6a_ch1Vz13R_pzYmQq5RV_cZpXoT0G7binTy-RJqkHq9wjD0kf88iN_9y28U2uQ' \
--data '{
        "alias": "browser-for-fun2",
        "description": "browser based authentication",
        "providerId": "basic-flow",
        "topLevel": true,
        "builtIn": true,
        "authenticationExecutions":[
            {
                "authenticator": "auth-cookie",
                "authenticatorFlow": false,
                "requirement": "ALTERNATIVE",
                "priority": 10,
                "autheticatorFlow": false,
                "userSetupAllowed": false
            },
            {
                "authenticator": "auth-spnego",
                "authenticatorFlow": false,
                "requirement": "DISABLED",
                "priority": 20,
                "autheticatorFlow": false,
                "userSetupAllowed": false
            },
            {
                "authenticator": "identity-provider-redirector",
                "authenticatorFlow": false,
                "requirement": "ALTERNATIVE",
                "priority": 25,
                "autheticatorFlow": false,
                "userSetupAllowed": false
            },
            {
                "authenticatorFlow": true,
                "requirement": "ALTERNATIVE",
                "priority": 30,
                "autheticatorFlow": true,
                "flowAlias": "Copy of browser forms",
                "userSetupAllowed": false
            }
        ]
}'`

using this curl , i can create flow but internal step are empty
![image](https://github.com/keycloak/keycloak/assets/57186072/cacd4d0d-f7fe-4d8e-b8c3-1e87a8980cb1)

**Comment by dteleguin** [comment]
Have you tried using `kcadm.[sh|bat]` tool instead of raw Admin REST API?

It could be that you'd need to create an empty flow first and then populate it with subflows and executors in subsequent calls.

**Reply by Bhavik237** [reply]
Hi @dteleguin , I have tried as you said
through API approach -

- create flow then added step It worked 

but there is one problem that I can add one step only at a time using this curl below
```

curl --location 'http://127.0.0.1:8080/admin/realms/heroes/authentication/flows/browser-for-fun/executions/execution' \
--header 'Content-Type: application/json' \
--data '{"provider":"reset-password"}' 
```

do not support multiple addition o step at the same time.

Kcadm.[sh|bat]- can share some documentation related to it  ,So I can try with that approach.

**Comment by lethargosapatheia** [comment]
Not only that, but you also cannot set the `requirement` parameter. It's just ignored. It works only in different step where you have to get the id (which is somewhat difficult to get consistently) and run `kcadm.sh update`.
